src/qibolab/_core/parameters.py

```diff
@@ -202,3 +223,121 @@ def replace(self, update: Update) -> "Parameters":
             _setvalue(d, path, val)
 
         return self.model_validate(d)
+
+
+QubitMap = dict[QubitId, Qubit]
+InstrumentMap = dict[InstrumentId, Instrument]
+
+
+class Hardware(Model):
+    """Part of the platform that specifies the hardware configuration."""
+
+    instruments: InstrumentMap
+    qubits: QubitMap
+    couplers: QubitMap = Field(default_factory=dict)
+
+
+def _gate_channel(qubit: Qubit, gate: str) -> str:
+    """Default channel that a native gate plays on."""
+    if gate in ("RX", "RX90", "CNOT"):
+        return qubit.drive
+    if gate == "RX12":
+        return qubit.drive_qudits[(1, 2)]
+    if gate == "MZ":
+        return qubit.acquisition
+    if gate in ("CP", "CZ", "iSWAP"):
+        return qubit.flux
+
+
+def _gate_sequence(qubit: Qubit, gate: str) -> Native:
+    """Default sequence corresponding to a native gate."""
+    channel = _gate_channel(qubit, gate)
+    pulse = Pulse(duration=0, amplitude=0, envelope=Rectangular())
+    if gate != "MZ":
+        return Native([(channel, pulse)])
+
+    return Native(
+        [(channel, Readout(acquisition=Acquisition(duration=0), probe=pulse))]
+    )
+
+
+def _pair_to_qubit(pair: str, qubits: QubitMap) -> Qubit:
+    """Get first qubit of a pair given in ``{q0}-{q1}`` format."""
+    q = tuple(pair.split("-"))[0]
+    try:
+        return qubits[q]
+    except KeyError:
+        return qubits[int(q)]
+
+
+def _native_builder(cls, qubit: Qubit, natives: set[str]) -> NativeContainer:
+    """Build default native gates for a given qubit or pair.
+
+    In case of pair, ``qubit`` is assumed to be the first qubit of the pair,
+    and a default pulse is added on that qubit, because at this stage we don't
+    know which qubit is the high frequency one.
+    """
+    return cls(
+        **{
+            gate: _gate_sequence(qubit, gate)
+            for gate in cls.model_fields.keys() & natives
+        }
+    )
+
+
+def _channel_config(id: ChannelId, channel: Channel) -> dict[ChannelId, Config]:
+    """Default configs correspondign to a channel."""
+    if isinstance(channel, DcChannel):
+        return {id: DcConfig(offset=0)}
+    if isinstance(channel, AcquisitionChannel):
+        return {id: AcquisitionConfig(delay=0, smearing=0)}
+    if isinstance(channel, IqChannel):
+        configs = {id: IqConfig(frequency=0)}
+        if channel.lo is not None:
+            configs[channel.lo] = OscillatorConfig(frequency=0, power=0)
+        if channel.mixer is not None:
+            configs[channel.mixer] = IqMixerConfig(frequency=0, power=0)
+        return configs
+    return {id: Config()}
+
+
+def initialize_parameters(
+    hardware: Hardware,
+    natives: Optional[set[str]] = None,
+    pairs: Optional[list[str]] = None,
+) -> Parameters:
+    """Generates default ``Parameters`` for a given hardware configuration."""
+    if natives is None:
+        natives = set()
+    else:
+        natives = set(natives)
```

Comment:
```suggestion
    natives = set(natives if natives is not None else ())
```
the usual ternary statement, since the is assignment is *not* conditional (only the value is conditioned).

Reply:
In principle:
- you could even make it even simpler, since `()` (empty tuple) is immutable, and thus an acceptable value for a default
    - still, `None` is a more natural default for an intentionally empty field
- given that the type hint is already `Optional[set[str]]`, if that's reliable, a set is already expected for a non-empty value - therefore, the following should be sufficient
    ```suggestion
        natives = natives if natives is not None else set()
    ```
    - that would be consistent, but if you want to be more lenient (that's not required - but it may be convenient if you already have a use case), I'd reflect that in the type hint

Reply:
Changed in fcbfddd0fa69d50bb357c3fc7535c3ea18ce2292, thanks.

Ideally, the type hint should be followed, however, if I remember correctly, this is a user facing function and since type hints are not strictly enforced in Python, we cannot be 100% certain it is reliable, that's why I think it is safer to cast to `set` internally. A way to enforce this was to convert the arguments to a `Model` class (the initial approach) and have pydantic enforce the type hints, but as discussed before this is not strictly required here.

Also, indeed, using empty tuples instead of `None` for this and also the `pairs` argument, could be another acceptable option. I left `None` as more natural.